src/cuda/poisson_fft.f90

```diff
@@ -0,0 +1,239 @@
+                                                    self%nz/2 + 1)
+
+      ! In-place forward FFT in x
+      ierrfft = cufftExecZ2Z(self%planZ2Zx, self%c_x_dev, self%c_x_dev, &
+                             CUFFT_FORWARD)
+
+   end subroutine fft_forward_cuda
+
+   subroutine fft_backward_cuda(self, f)
+      implicit none
+
+      class(cuda_poisson_fft_t) :: self
+      class(field_t), intent(inout) :: f
+
+      real(dp), device, pointer, dimension(:, :, :) :: f_dev
+      complex(dp), device, dimension(:, :, :), pointer :: c_x_ptr, c_y_ptr, &
+                                                          c_z_ptr
+
+      type(dim3) :: blocks, threads
+      integer :: ierrfft
+
+      select type(f); type is (cuda_field_t); f_dev => f%data_d; end select
+
+      ! In-place backward FFT in x
+      ierrfft = cufftExecZ2Z(self%planZ2Zx, self%c_x_dev, self%c_x_dev, &
+                             CUFFT_INVERSE)
+
+      ! Reorder from x to y
+      blocks = dim3(self%nx/SZ, self%ny/SZ, self%nz/2 + 1)
+      threads = dim3(SZ, SZ, 1)
+      c_x_ptr(1:self%nx, 1:SZ, 1:(self%ny*(self%nz/2 + 1))/SZ) => self%c_x_dev
+      c_y_ptr(1:self%ny, 1:SZ, 1:(self%nx*(self%nz/2 + 1))/SZ) => self%c_y_dev
+
+      call reorder_cmplx_x2y_T<<<blocks, threads>>>(c_y_ptr, c_x_ptr, &
+                                                    self%nz/2 + 1)
+
+      ! In-place backward FFT in y
+      ierrfft = cufftExecZ2Z(self%planZ2Zy, self%c_y_dev, self%c_y_dev, &
+                             CUFFT_INVERSE)
+
+      ! Reorder from y to z
+      blocks = dim3(self%nz/2/SZ + 1, self%ny/SZ, self%nx)
+      threads = dim3(SZ, SZ, 1)
+      c_y_ptr(1:self%ny, 1:SZ, 1:(self%nx*(self%nz/2 + 1))/SZ) => self%c_y_dev
+      c_z_ptr(1:self%nz/2 + 1, 1:SZ, 1:self%nx*self%ny/SZ) => self%c_z_dev
+
+      call reorder_cmplx_y2z_T<<<blocks, threads>>>(c_z_ptr, c_y_ptr, &
+                                                    self%nx, self%nz/2 + 1)
+
+      ! Backward FFT transform in z from complex to real
+      ierrfft = cufftExecZ2D(self%planZ2Dz, self%c_z_dev, self%f_tmp)
+
+      ! Finally reorder f back into our specialist data structure
+      blocks = dim3(self%nz/SZ, self%nx*self%ny/SZ, 1)
+      threads = dim3(SZ, SZ, 1)
+      call reshapeDSB<<<blocks, threads>>>(f_dev, self%f_tmp)
+
+   end subroutine fft_backward_cuda
+
+   subroutine fft_postprocess_cuda(self)
+      implicit none
+
+      class(cuda_poisson_fft_t) :: self
+
+      complex(dp), device, dimension(:, :, :), pointer :: c_dev, c_x_ptr
+      type(dim3) :: blocks, threads
+
+      ! Get some complex array pointers with right shape
+      c_x_ptr(1:self%nx, 1:SZ, 1:(self%ny*(self%nz/2 + 1))/SZ) => self%c_x_dev
+      c_dev(1:SZ, 1:self%nx, 1:(self%ny*(self%nz/2 + 1))/SZ) => self%c_y_dev
+
+      ! Reshape from cartesian-like to our specialist data structure
+      blocks = dim3(self%nx/SZ, (self%ny*(self%nz/2 + 1))/SZ, 1)
+      threads = dim3(SZ, SZ, 1)
```

Comment:
As previously commented for similar patterns

Reply:
Can be addressed as part of #73